### PR TITLE
[build] [test-suite] Move fake_ide to test-suite 

### LINKED ISSUE #14306: Since dune, make install has changed
`make install` installs files to a different layout now.  I assume the following changes are intended:
- `/usr/lib/coq{ => -core}/tools/{coqdoc/coqdoc.css,coqdoc/coqdoc.sty,CoqMakefile.in,TimeFileMaker.py,make-both-time-files.py,make-one-time-file.py,make-both-single-timing-files.py}`
- `/usr/lib/coq{ => -core}/{revision,META}`

I'm not sure about the following ones:
- `/usr/lib/coq/plugins/micromega/csdpcert` => `/usr/bin/csdpcert` (seems suspicious)
- `/usr/lib/coq/kernel/byterun/dllcoqrun.so` => `/usr/lib/stublibs/dllcoqrun_stubs.so` (maybe reasonable?)

The following changes are wrong:
- `/usr/share/texmf/tex/latex/misc/coqdoc.sty` => `/usr/share/coq-core/texmf/tex/latex/misc/coqdoc.sty`
- `/usr/bin/fake_ide` is now installed, whereas it wasn't previously (we might want to prefix this with `coq` and install it, but we definitely don't want to install it like this)

The following directories used to be installed but no longer are:
- `doc/stdlib/html`

Log of changes at https://launchpadlibrarian.net/537997360/buildlog_ubuntu-xenial-amd64.coq_8.master~git~202105101453+22306-0~daily347-c6223fc6aa~ubuntu16.04.1_BUILDING.txt.gz

### PR DESCRIPTION
**Pull request by Columbus240**
This PR fixes #14306 ~~in a very maintainable way~~.

Being only used for testing and nothing else, the `fake-ide` binary can go live in the test-suite/ directory. It'd be nice, if it fit into `test-suite/ide`, but I couldn't figure out how to do this. @ejgallego do you know how to do this? I don't know how to correctly write `test-suite/dune` and `test-suite/ide/dune` (if the latter is even necessary). Changing `./fake-ide.exe` to `ide/fake-ide.exe` in `test-suite/Makefile` would also be necessary.

Behavioral change: `fake-ide` now needs an additional cmd-line argument: the path where it can find `coqidetop`. Previously both binaries lived in the same folder, so this wasn't such a problem. Now this information has to be passed through by the Makefile.

Closes #14306
Succeeds #14611

### CODE REVIEW

File: lib/system.ml
Comment (SkySkimmer) [inline lib/system.ml:292]: The implicit argument doesn't seem to be used anywhere so can be removed.

**Review by SkySkimmer — COMMENTED** [review]
fake_ide doesn't seem to be built when the test suite is invoked outside dune so this won't work IIUC.

File: lib/system.ml
Comment (Columbus240) [reply lib/system.ml:292]: I don't understand what you mean. `byte` is used in an `if` clause and the unit-argument is necessary because the function wouldn't be callable otherwise (compilation error).

Comment (SkySkimmer) [reply lib/system.ml:292]: I mean it's not used by callers.

**Comment by Columbus240** [comment]
> fake_ide doesn't seem to be built when the test suite is invoked outside dune so this won't work IIUC.

It currently doesn't even work in dune. I made a mistake in `test-suite/dune`. Didn't notice it, because I didn't do `dune clean` before my final test.

File: lib/system.ml
Comment (Columbus240) [reply lib/system.ml]: Ah, ok. I misunderstood you. Have implemented your suggestion.

**Comment by Columbus240** [comment]
I don't know how to properly run the test-suite from a make-based build. I tried:
```
git clean -fdX
yes "" | ./configure
make test-suite
```
But it didn't work as expected. `ocamlfind` couldn't find "coq-core.toplevel" and the `coqc` tests couldn't find the plugins directory.
What is the correct invocation? Should something like the above even work?

File: lib/system.ml
Comment (gares) [inline lib/system.ml:294]: Since you touch this code, I suggest to use `Sys.win32 || Sys.cygwin`

File: test-suite/Makefile
Comment (gares) [inline test-suite/Makefile:693]: other parts of this makefile do stuff like `export coqc="eval $(coqc)"` to pass the path of coq binaries.

I think it is OK to pass the path of coqidetop to fake_ide, but I think you should pass the whole path, eg define a coqidetop variable in the Makefile and pass its value (akin to coqc). It should let you avoid to expose System.coq_executable_extension in the first place, IMO.

Comment (Columbus240) [reply test-suite/Makefile:693]: So fake_ide would read the environment variable?

**Comment by Zimmi48** [comment]
>`yes "" | ./configure`

Usually, developers using the make-based build do `./configure -profile devel`. That being said, running the test-suite with a non-dev configuration should also work.

**Comment by Columbus240** [comment]
Hm. No. `./configure -profile devel; make -f Makefile.build test-suite` gets me further, but still has some errors with finding plugins and other stuff.

File: test-suite/Makefile
Comment (gares) [reply test-suite/Makefile:693]: No, I was suggesting to have a Make variable contain the full path of coqidetop, like we do for coqc
https://github.com/coq/coq/blob/f684f34eed7889cbced249630e4ffe9306b1401b/test-suite/Makefile#L71
and pass that value to `./fake_ide.exe`. In that was you should not need to craft that path in fake_ide, you can just use the argument as is. This, in turn, will make it unnecessary to expose System.coq_executable_extension.

Comment (Columbus240) [reply test-suite/Makefile:693]: Ok, now I got it.

**Comment by gares** [comment]
from the logs (available on gilab as artifacts)
```
==========> TESTING ide/undo004.fake <==========
/bin/sh: 3: ./fake_ide.exe: not found
==========> FAILURE <==========
    ide/undo004.fake...Error!
```

**Comment by Columbus240** [comment]
Jup. Trying a different approach now. I take back everything about maintainability.
If `test-suite/Makefile` is called from a make-based build, it is run in `/test-suite`, but if it is called from `dune runtest`, it is run in `/_bin/default/test-suite/` and dune generates its output-files in `/_bin/default/`. This lead to the above error.
I wonder how many more revisions it takes. Ugh.

**Comment by gares** [comment]
Dear @coq/build-maintainers , do we really need the test suite to work with both build systems? Can't we settle that development happens with dune? (and test suite is for dev?)

**Comment by SkySkimmer** [comment]
I think we should drop the ability to call the test suite through dune (ie runtest), it is not very useful currently IMO
It's too early to drop compat with the legacy build though, we would have no way to test native for instance.

**Comment by Columbus240** [comment]
? `dune runtest` is the only way I can consistently & easily get the test-suite to run. I don't find it clearly documented, how to run the test-suite fro m make in a way that works.
`./configure -profile devel; make test-suite` doesn't seem to be correct.

**Comment by SkySkimmer** [comment]
`cd test-suite; make`

**Comment by Columbus240** [comment]
I tried the following, always after running `make clean; git clean -fdX; git clean -fd`.
* `cd test-suite; make`
* `./configure -profile devel; cd test-suite; make`
* `./configure -profile devel; make byte; cd test-suite; make`
None of them worked correctly.

**Comment by SkySkimmer** [comment]
`git clean -xffd; make -f Makefile.dune world; make -C test-suite`

**Comment by Columbus240** [comment]
Nope, still not working. I noticed that the [Mac CI](https://github.com/coq/coq/blob/master/.github/workflows/ci.yml) sets the `OCAMLPATH` variable before running the test-suite. My `OCAMLPATH` is empty. Maybe this has to do with it. Is yours empty as well?

**Comment by SkySkimmer** [comment]
Oh I thought https://github.com/coq/coq/pull/14625 got merged already

**Comment by Columbus240** [comment]
I managed to do the test-suite on a make-based build, by following the CI-instructions linked above. This time I tried the test-suite in both builds before pushing to GitHub.

**Comment by Columbus240** [comment]
I consider the PR ready to merge.
The fake_ide.exe is not registered & built by dune, but only appears in the Makefile. This way I could work around problems with constructing correct paths.

**Review by gares — APPROVED** [review]

**Comment by gares** [comment]
If nobody picks it before, I'll merge this on Monday.

**Comment by gares** [comment]
@coqbot merge now

**Pull request merged**
